sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java

```diff
@@ -307,18 +324,61 @@ public static Mono<BinaryData> fromFlux(Flux<ByteBuffer> data) {
      * @throws NullPointerException if {@code data} is null.
      */
     public static Mono<BinaryData> fromFlux(Flux<ByteBuffer> data, Long length) {
+        return fromFlux(data, length, true);
+    }
+
+    /**
+     * Creates an instance of {@link BinaryData} from the given {@link Flux} of {@link ByteBuffer}.
+     *
+     * <p><strong>Create an instance from a Flux of ByteBuffer</strong></p>
+     *
+     * <!-- src_embed com.azure.core.util.BinaryData.fromFlux#Flux-Long-boolean -->
+     * <pre>
+     * final byte[] data = &quot;Some Data&quot;.getBytes&#40;StandardCharsets.UTF_8&#41;;
+     * final long length = data.length;
+     * final boolean shouldAggregateData = false;
+     * final Flux&lt;ByteBuffer&gt; dataFlux = Flux.just&#40;ByteBuffer.wrap&#40;data&#41;&#41;;
+     *
+     * Mono&lt;BinaryData&gt; binaryDataMono = BinaryData.fromFlux&#40;dataFlux, length, shouldAggregateData&#41;;
+     *
+     * Disposable subscriber = binaryDataMono
+     *     .map&#40;binaryData -&gt; &#123;
+     *         System.out.println&#40;binaryData.toString&#40;&#41;&#41;;
+     *         return true;
+     *     &#125;&#41;
+     *     .subscribe&#40;&#41;;
+     *
+     * &#47;&#47; So that your program wait for above subscribe to complete.
+     * TimeUnit.SECONDS.sleep&#40;5&#41;;
+     * subscriber.dispose&#40;&#41;;
+     * </pre>
+     * <!-- end com.azure.core.util.BinaryData.fromFlux#Flux-Long-boolean -->
+     *
+     * @param data The {@link Flux} of {@link ByteBuffer} that {@link BinaryData} will represent.
+     * @param length The length of {@code data} in bytes.
+     * @param aggregate A flag indicating whether {@link Flux} should be aggregated (buffered) eagerly or
+     *                  consumption deferred.
+     * @return A {@link Mono} of {@link BinaryData} representing the {@link Flux} of {@link ByteBuffer}.
+     * @throws IllegalArgumentException if the length is less than zero.
+     * @throws NullPointerException if {@code data} is null.
+     */
+    public static Mono<BinaryData> fromFlux(Flux<ByteBuffer> data, Long length, boolean aggregate) {
         if (data == null) {
             return monoError(LOGGER, new NullPointerException("'content' cannot be null."));
         }
         if (length != null && length < 0) {
             return monoError(LOGGER, new IllegalArgumentException("'length' cannot be less than 0."));
         }
-        if (length != null) {
-            return FluxUtil.collectBytesInByteBufferStream(data, length.intValue())
+        if (aggregate) {
+            if (length != null) {
```

Comment:
Let's eagerly `OutOfMemoryError` here is `length` > `Integer.MAX_VALUE` (and I think most JVM implementations OOME at `Integer.MAX_VALUE` - 8).

Or, if we want, add an overload to the collector utility method that takes `long` and throws if it is over the limit. And last thought, we can do better by providing an implementation only version of `BinaryData` that is based on `List<byte[]>` so we can support >2GB, similar to `SequenceInputStream` in concept.

Reply:
Something like https://github.com/Azure/azure-sdk-for-java/blob/3f31d68eed6fbe11516ca3afe3955c8840a6e974/sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/BufferAggregator.java .

This reminds me to add one more validation rule.

Reply:
https://github.com/Azure/azure-sdk-for-java/issues/28799